XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs

```diff
@@ -340,14 +377,43 @@ void client_DownloadFileCompleted(object sender, AsyncCompletedEventArgs e)
             log.DebugFormat("XenServer patch '{0}' download completed successfully", updateName);
         }
 
-        public override void RecomputeCanCancel()
+        private void NetworkAvailabilityChanged(object sender, NetworkAvailabilityEventArgs e)
         {
-            CanCancel = !Cancelling && !IsCompleted && (patchDownloadState == DownloadState.InProgress);
+            if (!e.IsAvailable && client != null && patchDownloadState == DownloadState.InProgress)
+            {
+                patchDownloadError = new WebException(Messages.NETWORK_CONNECTIVITY_ERROR);
+                patchDownloadState = DownloadState.Error;
+                client.CancelAsync();
+            }
         }
+    }
 
-        protected override void CancelRelatedTask()
+
+    public class UnzipXenServerPatchAction : DownloadAndUnzipXenServerPatchActionBase
+    {
+        private readonly string _zippedUpdatePath;
+
+        public UnzipXenServerPatchAction(string zippedUpdatePath, params string[] updateFileExtensions)
+            : base(Path.GetFileNameWithoutExtension(zippedUpdatePath), updateFileExtensions)
+        {
+            _zippedUpdatePath = zippedUpdatePath;
+            Title = string.Format(Messages.UPDATES_WIZARD_EXTRACT_ACTION_TITLE, Path.GetFileNameWithoutExtension(zippedUpdatePath));
+        }
+
+        protected override void Run()
         {
+            PatchPath = ExtractFile(_zippedUpdatePath, false);
+            Description = Messages.COMPLETED;
+        }
+
+        protected override void UpdateExtractionProgress(long bytesTransferred, long totalBytesToTransfer)
+        {
+            PercentComplete = (int)(100.0 * bytesTransferred / totalBytesToTransfer);
         }
 
+        public override void RecomputeCanCancel()
+        {
+            CanCancel = !Cancelling && !IsCompleted;

```

Comment:
again not sure of the point of this function that doesn't get called or assign anything that is used

Reply:
Same as above, it's base class infrastructure.